File: client/src/types/order.ts
/**
 * @file order.ts
 * @description 定义订单相关的数据类型，包括订单项、配送方式等
 * @input 无
 * @output OrderItem, OrderItemProduct, AddressInfo, PlacedOrder, UserOrder 等类型及常量
 */

/**
 * 订单项中的商品信息
 */
export interface OrderItemProduct {
  /** 商品 ID */
  id: string
  /** 商品名称 */
  name: string
  /** 商品单价 */
  price: number
  /** 商品图片 URL 列表 */
  imageUrls: string[]
}

/**
 * 订单项，表示订单中的单个商品条目
 */
export interface OrderItem {
  /** 订单项 ID */
  id: string
  /** 购买数量 */
  amount: number
  /** 订单项总价 */
  price: number
  /** 关联的商品信息 */
  product: OrderItemProduct
}

/**
 * 收货地址信息
 */
export interface AddressInfo {
  /** 收货人姓名 */
  receiverName: string
  /** 收货人联系电话 */
  receiverPhone: string
  /** 省/市/区 */
  region: string
  /** 详细地址 */
  detailAddress: string
}

/**
 * 已提交的订单数据结构
 */
export interface PlacedOrder {
  /** 订单 ID */
  id: string
  /** 订单总价 */
  price: number
  /** 订单状态 */
  status: string
  /** 收货人姓名 */
  receiverName: string
  /** 收货人联系电话 */
  receiverPhone: string
  /** 省/市/区 */
  region: string
  /** 详细地址 */
  detailAddress: string
  /** 配送方式 */
  deliveryMethod: string
  /** 支付方式 */
  paymentMethod: string
  /** 下单时间 */
  received: string
  /** 预计送达时间，可能为空 */
  expectedDelivery: string | null
  /** 订单商品列表 */
  items: OrderItem[]
}

/**
 * 管理员视角的用户订单数据结构，包含用户信息
 */
export interface UserOrder {
  /** 订单 ID */
  id: string
  /** 用户邮箱 */
  userEmail: string
  /** 用户名 */
  username: string
  /** 订单总价 */
  price: number
  /** 订单状态 */
  status: string
  /** 下单时间 */
  received: string
  /** 预计送达时间，可能为空 */
  expectedDelivery: string | null
  /** 订单商品列表 */
  items: OrderItem[]
}

/**
 * 订单状态枚举
 */
export enum OrderStatus {
  /** 待处理 */
  PENDING,
  /** 已发货 */
  SHIPPED,
  /** 已送达 */
  DELIVERED
}

/**
 * 将订单状态枚举转换为字符串
 * @param {OrderStatus} status - 订单状态枚举值
 * @returns {string} 状态对应的字符串表示
 */
export function orderStatusToString(status: OrderStatus): string {
  switch (status) {
    case OrderStatus.PENDING:
      // 待处理状态
      return 'PENDING'
    case OrderStatus.SHIPPED:
      // 已发货状态
      return 'SHIPPED'
    case OrderStatus.DELIVERED:
      // 已送达状态
      return 'DELIVERED'
    default:
      // 未知状态
      return 'INVALID_STATUS'
  }
}

/** 配送方式映射：英文键 -> 中文显示名 */
export const DELIVERY_METHODS: Record<string, string> = {
  STANDARD_DELIVERY: '普通快递',
  EXPRESS_DELIVERY: '加急快递'
}

File: server/src/main/java/me/code/springboot_postgres/models/entities/Order.java
/**
 * @file Order.java
 * @description 订单实体类，表示用户的购买订单
 * @input 用户、订单项列表、收货人信息、配送方式
 * @output 持久化的订单记录
 */
package me.code.springboot_postgres.models.entities;

import jakarta.persistence.*;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.UuidGenerator;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.time.LocalDateTime;
import java.util.List;

/**
 * 订单实体
 * 职责：映射用户订单数据，包含订单状态、价格、配送信息和订单项列表
 */
@Getter
@Setter
@NoArgsConstructor
@Entity
@Table(name = "orders",
        indexes = {
            @Index(name = "idx_orders_status", columnList = "status")
        })
public class Order {

    // 订单唯一标识
    @Id
    @GeneratedValue
    @UuidGenerator
    private String id;

    // 订单状态
    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 15)
    private Status status;

    // 订单总价格
    @Column(nullable = false, precision = 10, scale = 2)
    private BigDecimal price;

    // 支付方式
    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 25)
    private PaymentMethod paymentMethod;

    // 收货人姓名
    @Column(nullable = false, length = 50)
    private String receiverName;

    // 联系电话
    @Column(nullable = false, length = 20)
    private String receiverPhone;

    // 省/市/区
    @Column(nullable = false, length = 100)
    private String region;

    // 详细地址
    @Column(nullable = false, columnDefinition = "TEXT")
    private String detailAddress;

    // 配送方式
    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 30)
    private DeliveryMethod deliveryMethod;

    // 订单接收时间
    private LocalDateTime received;

    // 预计送达时间
    @Column(name = "expected_delivery")
    private LocalDateTime expectedDelivery;

    // 下单用户（懒加载）
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    // 订单项列表（级联保存、懒加载、孤立删除）
    @OneToMany(mappedBy = "order", cascade = CascadeType.ALL, fetch = FetchType.LAZY, orphanRemoval = true)
    private List<OrderItem> items = new java.util.ArrayList<>();

    /**
     * 构造订单对象，自动计算总价并设置初始状态
     * 支付方式固定为余额支付
     * @param user 下单用户
     * @param items 订单项列表
     * @param receiverName 收货人姓名
     * @param receiverPhone 联系电话
     * @param region 省/市/区
     * @param detailAddress 详细地址
     * @param deliveryMethod 配送方式
     */
    public Order(User user, List<OrderItem> items, String receiverName, String receiverPhone,
                 String region, String detailAddress, DeliveryMethod deliveryMethod) {
        this.status = Status.PENDING;
        this.user = user;
        this.items = items;
        this.price = getTotalPrice();
        this.receiverName = receiverName;
        this.receiverPhone = receiverPhone;
        this.region = region;
        this.detailAddress = detailAddress;
        this.deliveryMethod = deliveryMethod;
        // 支付方式固定为余额支付
        this.paymentMethod = PaymentMethod.ACCOUNT_BALANCE;
        this.received = LocalDateTime.now();
        this.expectedDelivery = null;
    }

    /**
     * 计算订单总价（所有订单项价格之和，保留两位小数）
     * @return 订单总价
     */
    public BigDecimal getTotalPrice() {
        return items.stream()
                .map(OrderItem::getPrice)
                .reduce(BigDecimal.ZERO, BigDecimal::add)
                .setScale(2, RoundingMode.HALF_UP);
    }

    /**
     * 订单状态枚举
     */
    public enum Status {
        PENDING, SHIPPED, DELIVERED
    }

    /**
     * 配送方式枚举
     */
    public enum DeliveryMethod {
        STANDARD_DELIVERY, EXPRESS_DELIVERY
    }

    /**
     * 支付方式枚举（仅支持余额支付）
     */
    public enum PaymentMethod {
        ACCOUNT_BALANCE
    }
}
